<a href="https://colab.research.google.com/github/PeggyK1/DS-Unit-1-Sprint-2-Statistics/blob/master/Peggy_Krom_DS_124_Introduction_to_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science Module 124

## Introduction to Bayesian Inference




## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):` 

You should only truly need these three values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?


In [0]:
# TODO - code!

In [0]:
import numpy as np

In [0]:
#True Positive Rate
p_true_pos = 1
#Prior Belief of Drunk Drivers
p_dd = 1/1000
#False Positive Rate
p_false_pos = .08
#Non Drunk Driver
p_non_dd = 1-p_dd


In [0]:
#While Loop Function
def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):
  numerator = true_positive_rate*prob_drunk_prior
  denominator = (true_positive_rate*prob_drunk_prior) + (false_positive_rate*(1-prob_drunk_prior))
  posterior_probability = numerator / denominator
  return posterior_probability

x=.001
while x < .95:
  x = prob_drunk_given_positive(x, .08, 1)
  print(x)


0.012357884330202669
0.13525210993291495
0.6615996951348605
0.9606895076105054


In [0]:
#Messy Function
def drunk_drivers():
  p_true_pos = 1
  p_dd= 1/1000
  p_false_pos = .08
  p_non_dd = 1-p_dd
  numerator = p_true_pos*p_dd
  denominator = p_true_pos*p_dd + p_false_pos*p_non_dd
  result=numerator/denominator

  return(result)


In [0]:
drunk_drivers

<function __main__.drunk_drivers>

In [0]:
post_prob = drunk_drivers()
post_prob

0.012357884330202669

In [0]:
def drunk_drivers():
  p_true_pos = 1
  p_dd= post_prob
  p_false_pos = .08
  p_non_dd = 1-p_dd
  numerator = p_true_pos*p_dd
  denominator = p_true_pos*p_dd + p_false_pos*p_non_dd
  result=numerator/denominator

  return(result)

In [0]:
post_prob = drunk_drivers()
post_prob

0.13525210993291495

In [0]:
post_prob = drunk_drivers()
post_prob

0.6615996951348605

In [0]:
post_prob = drunk_drivers()
post_prob

0.9606895076105054

In [0]:
post_prob = drunk_drivers()
post_prob

0.9967371577896734

In [0]:
post_prob = drunk_drivers()
post_prob

0.9997381867081508

In [0]:
post_prob = drunk_drivers()
post_prob

0.9999790498904363

**After 1 Test:** The probability is 1.2%
**After 2 Tests:** The probability is 13.5%
**After 3 Tests:** The probability is 66%
**After 4 Tests:** The probability is 96%

It took 4 tests to get the probability over 95%

It just keeps increasing with each test and the probability gets closer and closer to 100%

### 2) In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/).

Bayesian statistics is based on probability. The probability is continuously updated as more results are calculated and more information is found. The biggest weakness is that the prior probabilities are subjective.

Frequentist statistics uses samples of populations. It compares the samples, including confidence levels and margins of error, to estimate how close the sample values are to the the true population values. The biggest weakness is that it is limited by sample size, small samples do not work.

### 3) Use the following Template to help come up with ideas for your Build Sprint Project: 



---

## Idea 1: Medical Turnover Rates

### You
**What do you care about?** What is causing higher turnover rates in the medical field

**What do you know about?** The turnover rate in the medical field is rising for burnout and other reasons.

**What decisions do you face?** Are there enough connections in the data to answer questions about turnover rates.
Is the data recent enough to provide relevant results




### Seven templates
Training Kit (https://learn.lambdaschool.com/ds/module/recedjanlbpqxic2r) explains the seven templates from Priceonomics.

Can you apply the templates to your topics?

**Geographic Variation:** Yes

**Trend related to the news:** No

**Who does that?:** No

**Answering a question people care about:** Yes

**Valuable to businesses:** No

**What's the most popular?:** No

**Cost/Money rankings:** No

I could look at reasons for turnover in different regions and see if there are patterns between geographical locations.
I can look at patterns in the reasons and patterns in the solutions that have been tried to mitigate the turnover rates and see if the data can answer questions about them. 

### Misconceptions

What misconceptions do people have about your topic?

The vacancies are just from "babyboomers" retiring.

### Examples

What data storytelling example inspires you?

Could you do a new hypothesis, for the same question?

Could you do a new question, for the same topic?

Could you do a new topic, with the same "style"?

###Data

Where could you search for data about your topic?

datasetsearch.research.google.com

---

## Idea 2: Relationships Between Popular Board Games

### You
**What do you care about?** Is a certain type of board game more popular

**What do you know about?** There are several different rankings of board games. There are different types of board game lovers, some like short simple games others like longer more complex games.

**What decisions do you face?** Are there enough connections in the data to form a conclusion or is it too varied



### Seven templates
Training Kit (https://learn.lambdaschool.com/ds/module/recedjanlbpqxic2r) explains the seven templates from Priceonomics.

Can you apply the templates to your topics?

**Geographic Variation:** Yes

**Trend related to the news:** No

**Who does that?:** Yes

**Answering a question people care about:** No

**Valuable to businesses:** No

**What's the most popular?:** Yes

**Cost/Money rankings:** Yes

I could see if there are patterns in popularity between different geographical locations or age groups. Does cost play a part in most popular.

### Misconceptions

What misconceptions do people have about your topic?

"Only nerds like board games"

### Examples

What data storytelling example inspires you? "How Much of Your Life the US has Been at War"

Could you do a new hypothesis, for the same question? No

Could you do a new question, for the same topic? Birthyear and the percentage of population born each year that was involved in war

Could you do a new topic, with the same "style"? Yes

###Data

Where could you search for data about your topic?

Kaggle

## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 & 3 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!